### PR TITLE
fix possible sensitive information leak in flare/workloadlist

### PR DESCRIPTION
**Pull request by zhuminyi**
### What does this PR do?
Fix possible sensitive information leak in agent workloadlist and flare workloadlist.log when information is added to annotation in plaintext. 
Bug is reported in this [incident](https://app.datadoghq.com/incidents/22663)

ScrubJSON function is a similar to ScrubYAML which converts string to data obj and scrubs the crediential recursively. The function will be called when 1) workloadlist command is triggered and 2)verbose option is chosen. There will be no impact on overall performance of other command or any other scrubbing util functions.

### Possible Drawbacks / Trade-offs

When we convert string to json data, the original field order is not preserved. So after scrubbing and convert it back to string, the order of new annotation field might be different from original one.  For example, origin order in the displayed string is [“host”: “%%host%%“, “port” : 5432, “username”: “postgresadmin”, “password”: “admin1234”], new order might be [“host”: “%%host%%“, “port” : 5432,  “password”: “********“, “username”: “postgresadmin”].

### Describe how to test/QA your changes
Add plaintext credential to config, for example 

>      annotations:
>         ad.datadoghq.com/nginx.checks: |
>           {
>             "nginx": {
>               "init_config": {},
>               "instances": [
>                 {
>                   "nginx_status_url": "http://%%host%%:8080/nginx_status",
>                    "username": "admin", 
>                    "password": "admin123"
>                 }
>               ]
>             }
>           }

run `agent workloadlist -v` or `agent flare`, check the workloadlist output, the password field value should be replaced by "********"

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/flare/archive_test.go [hunk 1]
```diff
@@ -166,6 +167,7 @@ func TestGetWorkloadList(t *testing.T) {
 	assert.Contains(t, string(content), "Name: init-volume ID: e19e1ba787")
 	assert.Contains(t, string(content), "container_id_2")
 	assert.Contains(t, string(content), "Name: init-config ID: 4e0ffee5d6")
+	assert.NotContains(t, string(content), "Name: init-passwd ID: f17539af8d pwd: admin123")
```
Comment (dustmop) [inline pkg/flare/archive_test.go:170]: What do you think about changing from `NotContains(...)` to `Contains(... "pwd: *********")` to show the actual output of the scrub?

File: pkg/workloadmeta/dump.go [hunk 2]
```diff
@@ -33,6 +34,9 @@ func (wdr WorkloadDumpResponse) Write(writer io.Writer) {
 
 	for kind, entities := range wdr.Entities {
 		for entity, info := range entities.Infos {
+			info = scrubMessage(info)
```
Comment (hush-hush) [inline pkg/workloadmeta/dump.go:37]: I don't think this solves the issue of sensitive data in a JSON blob from docker label. Or am I missing something ?

File: pkg/flare/archive_test.go (hunk 1 above)
Comment (zhuminyi) [reply pkg/flare/archive_test.go]: right, updated.

File: pkg/util/scrubber/default.go [hunk 3]
```diff
@@ -283,6 +283,22 @@ func ScrubYamlString(data string) (string, error) {
...
+// ScrubJSON scrubs credentials from the given JSON by loading the data and scrubbing the object instead of the
+// serialized string, using the default scrubber.
+func ScrubJSON(data []byte) ([]byte, error) {
+	return DefaultScrubber.ScrubJSON(data)
+}
+
+// ScrubJSONString scrubs credentials from the given JSON string by loading the data and scrubbing the object instead of
+// the serialized string, using the default scrubber.
+func ScrubJSONString(data string) (string, error) {
+	res, err := DefaultScrubber.ScrubJSON([]byte(data))
```
Comment (clamoriniere) [inline pkg/util/scrubber/default.go:295]: should it call `ScrubJSON()` instead ? 
```suggestion
	res, err := ScrubJSON([]byte(data))
```

File: pkg/util/scrubber/test/config.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{
+    "api_key": "aaaaaaaaaaaaaaaaaaaaaaaaaaaabbbb",
+    "api_key2": "aaaaaaaaaaaaaaaaaaaaaaaaaaaabbbb",
+    "api_key3": "aaaaaaaaaaaaaaaaaaaaaaaaaaaabbbb",
+    "app_key": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabbbb",
+    "app_key2": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabbbb",
+    "app_key3": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabbbb",
+    "password": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabbbb",
```
Comment (clamoriniere) [inline pkg/util/scrubber/test/config.json:8]: should we add also `pwd` key ?

File: pkg/workloadmeta/types.go [hunk 5]
```diff
@@ -295,7 +295,7 @@ func (e EntityMeta) String(verbose bool) string {
 	_, _ = fmt.Fprintln(&sb, "Namespace:", e.Namespace)
 
 	if verbose {
-		_, _ = fmt.Fprintln(&sb, "Annotations:", mapToString(e.Annotations))
+		_, _ = fmt.Fprintln(&sb, "Annotations:", mapToScrubbedJSONString(e.Annotations))
```
Comment (AliDatadog) [inline pkg/workloadmeta/types.go:298]: Is this the only place where a secret can be leaked ? I think container labels might also be affected.

**Review by clamoriniere — APPROVED** [review]

File: pkg/util/scrubber/json_scrubber.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		return
+	}
+	switch v := (*data).(type) {
+	case map[string]interface{}:
+		walkJSONInterface(v, callback)
+	case []interface{}:
+		walkJSONSlice(v, callback)
+	}
+}
+
```
Comment (hush-hush) [inline pkg/util/scrubber/json_scrubber.go:53]: This looks similar to the YAML version, could we not reuse it ?

**Review by AliDatadog — APPROVED** [review]
Approved but I think we might need a solution that works with any kind of string. For example, I think ENV variables could leak in the history of a container image:
example
```
agent workload-list -v
[...]
Media Type:
Digest:
Size in bytes: 0
URLs: []
History:
- createdAt: 2023-09-27 11:48:29.659722591 +0000 UTC
- createdBy: /bin/sh -c #(nop)  ENV PASSWORD=123
- comment:
- emptyLayer: true
```

**Review by hush-hush — APPROVED** [review]

**Pull request merged**